bokeh/application/handlers/directory.py

```diff
@@ -120,14 +121,30 @@ def __init__(self, *args, **kwargs):
         handler = NotebookHandler if main.endswith('.ipynb') else ScriptHandler
         self._main_handler = handler(filename=self._main, argv=argv, package=self._package)
 
+        hooks = None
+        app_hooks = join(src_path, 'app_hooks.py')
         lifecycle = join(src_path, 'server_lifecycle.py')
-        if exists(lifecycle):
-            self._lifecycle = lifecycle
+        if exists(app_hooks) and exists(lifecycle):
+            raise ValueError("Directory style apps can provide either server_lifecycle.py or app_hooks.py, not both.")
+        elif exists(lifecycle):
+            hooks = lifecycle
+        elif exists(app_hooks):
+            hooks = app_hooks
+
+        if hooks is not None:
+            self._lifecycle = hooks
             self._lifecycle_handler = ServerLifecycleHandler(filename=self._lifecycle, argv=argv, package=self._package)
         else:
             self._lifecycle = None
             self._lifecycle_handler = Handler() # no-op handler
 
+        if exists(app_hooks):
+            self._request_handler = hooks
+            self._request_handler = ServerRequestHandler(filename=self._request_handler, argv=argv, package=self._package)
+        else:
+            self._request_handler = None
+            self._request_handler = Handler() # no-op handler
+
```

Comment:
@bryevdv, @philippjfr, can you explain this logic?  Right now if `lifecycle` doesn't exist, then a handler for it will be configured with `hooks = app_hooks` regardless.

Reply:
https://docs.bokeh.org/en/latest/docs/releases.html?highlight=app_hooks#server-lifecycle-hooks

Reply:
Yeah, those double assignments are weird, trying to figure out if that first assignment should actually be assigning to something else. Looks like those release notes also need an edit:

> Server lifecycle hooks **should now goin in a file** named app_hooks.py, which is also where new request hooks are to be defined.

Reply:
Is app_hooks actually what I should use to mock and modify defaults of Panel. C.f https://github.com/holoviz/panel/issues/2323